test/parse/recognizer_spec.ts

```diff
@@ -1529,6 +1529,59 @@ function defineRecognizerSpecs(
                 expect(serializedGrammar[1].type).to.equal("Rule")
             })
 
+            it("can use serialized grammar in performSelfAnalysis", () => {
+                let serializedGrammar = null
+                class SerializingParser extends Parser {
+                    constructor(input: IToken[] = []) {
+                        super(input, [PlusTok, MinusTok, IdentTok])
+                        this.performSelfAnalysis(serializedGrammar)
+                    }
+
+                    public rule = this.RULE("rule", () => {
+                        this.AT_LEAST_ONE_SEP({
+                            SEP: IdentTok,
+                            DEF: () => {
+                                this.SUBRULE(this.rule2)
+                            }
+                        })
+                    })
+
+                    public rule2 = this.RULE("rule2", () => {
+                        this.OR([
+                            {
+                                ALT: () => {
+                                    this.CONSUME1(MinusTok)
+                                }
+                            },
+                            {
+                                ALT: () => {
+                                    this.CONSUME1(PlusTok)
+                                }
+                            }
+                        ])
+                    })
+                    // a rule to exercise certain cases in deserializeGrammar
+                    public rule3 = this.RULE("rule3", () => {
+                        this.AT_LEAST_ONE(() => {
+                            this.CONSUME1(IdentTok)
+                        })
+                        this.MANY_SEP({
+                            SEP: PlusTok,
+                            DEF: () => {
+                                this.CONSUME2(IdentTok)
+                            }
+                        })
+                    })
+                }
+                let parser = new SerializingParser([])
+                serializedGrammar = parser.getSerializedGastProductions()
+                clearCache()
+                let parser1 = new SerializingParser([])
+                expect(parser.getGAstProductions()).to.deep.equal(
```

Comment:
this assertion does not really check anything because both for parser1 and parser it would
access the shared cache.

Reply:
Isn’t that what the `clearCache()` call on line 1578 prevents? I initially hadn’t cleared the cache and code coverage tool warned that the deserialize logic wasn’t being exercised.

Reply:
After clearing the cache the de-serialization will work once so that solved the coverage.
However calling getGastProductions on either parser afterwards would return the same
Object from the same global cache (maybe not exactly the same as it might be cloned).

Just debug the call to  getGAstProductions to see...

You want to save the gastProductions **before** clearing the cache to compare them
to after clearing + de-serialization.

Reply:
🤦‍♂️ makes sense!